src/dynamics/ergodic/add_circle.lean

```diff
@@ -0,0 +1,103 @@
+/-
+Copyright (c) 2022 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import measure_theory.group.add_circle
+import dynamics.ergodic.ergodic
+import measure_theory.covering.density_theorem
+import data.set.pointwise.iterate
+
+/-!
+# Ergodic maps of the additive circle
+
+This file contains proofs of ergodicity for maps of the additive circle.
+
+# Main definitions:
+
+ * `add_circle.ergodic_zsmul`: given `n : ℤ` such that `1 < |n|`, the self map `y ↦ n • y` on
+   the additive circle is ergodic (wrt the Haar measure).
+ * `add_circle.ergodic_nsmul`: given `n : ℕ` such that `1 < n`, the self map `y ↦ n • y` on
+   the additive circle is ergodic (wrt the Haar measure).
+
+# TODO
+
+ * Show that the map `y ↦ n • y + x` is ergodic for any `x` (and `1 < |n|`).
+
+-/
+
+open set function measure_theory measure_theory.measure filter metric
+open_locale measure_theory nnreal ennreal topological_space pointwise
+
+namespace add_circle
+
+variables {T : ℝ} [hT : fact (0 < T)]
+include hT
+
+/-- If a null-measurable subset `s` of the circle is invariant under rotation by a rational angle
+with arbitarily large denominator, then it must be almost empty or almost full. -/
+lemma ae_empty_or_univ_of_forall_vadd_eq_self
+  {s : set $ add_circle T} (hs : null_measurable_set s volume)
+  {ι : Type*} {l : filter ι} [l.ne_bot] {u : ι → add_circle T}
+  (hu₁ : ∀ i, (u i) +ᵥ s = s) (hu₂ : tendsto (add_order_of ∘ u) l at_top) :
+  s =ᵐ[volume] (∅ : set $ add_circle T) ∨ s =ᵐ[volume] univ :=
```

Comment:
I think my confusion stems from trying to work out which variable is "the" angle; at a glance, it looks like you actually have a family of angles `u i`.

Reply:
Ah OK. I have reworded the comment to say "family of rational angles" which I hope will help with this.